component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/OneTimePassword.java

```diff
@@ -0,0 +1,248 @@
+     *                         should be appended to the OTP.
+     * @param truncationOffset the offset into the MAC result to begin truncation. If this
+     *                         value is out of the range of 0 ... 15, then dynamic truncation
+     *                         will be used. Dynamic truncation is when the last 4 bits of
+     *                         the last byte of the MAC are used to determine the start
+     *                         offset.
+     * @throws NoSuchAlgorithmException if no provider makes either HmacSHA1 or HMAC-SHA-1 digest
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-1 key.
+     */
+    public static String generateAlphaNumericOTP(byte[] secret, long movingFactor,
+                                                 int codeDigits, boolean addChecksum,
+                                                 int truncationOffset)
+            throws NoSuchAlgorithmException, InvalidKeyException {
+        // put movingFactor value into text byte array
+        String result = null;
+        int digits = addChecksum ? (codeDigits + 1) : codeDigits;
+        byte[] text = new byte[8];
+        for (int i = text.length - 1; i >= 0; i--) {
+            text[i] = (byte) (movingFactor & 0xff);
+            movingFactor >>= 8;
+        }
+        // compute hmac hash
+        byte[] hash = hmacShaGenerate(secret, text);
+        // put selected bytes into result int
+        int offset = hash[hash.length - 1] & 0xf;
+        if ((0 <= truncationOffset) && (truncationOffset < (hash.length - 8))) {
+            offset = truncationOffset;
+        }
+        int firstBinary = ((hash[offset] & 0x7f) << 24) | ((hash[offset + 1] & 0xff) << 16)
+                | ((hash[offset + 2] & 0xff) << 8) | ((hash[offset + 3] & 0xff));
+        int secondBinary = ((hash[offset + 4] & 0x7f) << 24) | ((hash[offset + 5] & 0xff) << 16)
+                | ((hash[offset + 6] & 0xff) << 8) | ((hash[offset + 7] & 0xff));
+        result = Integer.toString(firstBinary, 36).concat(Integer.toString(secondBinary, 36)).toUpperCase();
+        while (result.length() < digits) {
+            result = "A" + result;
+        }
+        result = result.substring(result.length() - digits, result.length());
+        return result;
+    }
+
+    /**
+     * Generate the token.
+     *
+     * @param key                       the key
+     * @param base                      the base
+     * @param digits                    the number of digits
+     * @param isEnableAlphanumericToken a flag that indicates the token is alphanumeric or not
+     * @return the generated token
+     */
+    public String generateToken(String key, String base,
+                                int digits, boolean isEnableAlphanumericToken) throws AuthenticationFailedException {
+
+        int truncOffset = 0;
+        if (isEnableAlphanumericToken) {
+            try {
```

Comment:
Shall we use a single try catch block to avoid code duplication.

ex:

public String generateToken(String key, String base,
                            int digits, boolean isEnableAlphanumericToken) throws AuthenticationFailedException {

    int truncOffset = 0;
    try {
         if (isEnableAlphanumericToken) {
            return generateAlphaNumericOTP(key.getBytes(), Long.parseLong(base), digits, false, truncOffset);
        } else {
            return generateOTP(key.getBytes(), Long.parseLong(base), digits, false, truncOffset);
        }
    } catch (NoSuchAlgorithmException e) {
        throw new AuthenticationFailedException("Unable to find the SHA1 Algorithm to generate OTP", e);
    } catch (InvalidKeyException e) {
        throw new AuthenticationFailedException("Unable to find the secret key", e);
    }
}